File: CodeZen-main/limit_questions.py

import re
import os

file_path = r'c:\Users\lalev\Downloads\CodeZen-main\CodeZen-main\lib\data.ts'

with open(file_path, 'r', encoding='utf-8') as f:
    content = f.read()

# Pattern to find the test array for each course
# It looks for test: [ ... ]
# We want to find each occurrence and limit the items inside.

def limit_questions(match):
    test_content = match.group(1)
    # Split by }, which separates objects
    # Note: there might be nested objects but in this schema TestQuestion is simple.
    # Actually, let's use a more robust way to find individual questions.
    # Questions start with { and end with }
    
    questions = re.findall(r'\{\s*id:[\s\S]*?\}', test_content)
    
    if len(questions) > 10:
        limited_questions = questions[:10]
        # Join them back
        # We need to preserve some formatting if possible, but joining with },\n is a start.
        new_test_content = ',\n      '.join(limited_questions)
        return f'test: [\n      {new_test_content}\n    ],'
    return match.group(0)

# This regex might be too simple, let's refine it.
# We want to match: test: [ <anything> ],
# We use a non-greedy match for the content until the first ], that follows a }
new_content = re.sub(r'test:\s*\[([\s\S]*?)\]\s*,', limit_questions, content)

with open(file_path, 'w', encoding='utf-8') as f:
    f.write(new_content)

print("Processed and limited questions to 10 per course.")

File: CodeZen-main/limit_questions.js

const fs = require('fs');
const path = require('path');

const filePath = path.join('C:', 'Users', 'lalev', 'Downloads', 'CodeZen-main', 'CodeZen-main', 'lib', 'data.ts');

// Re-read the file (it might be partially broken now)
let content = fs.readFileSync(filePath, 'utf8');

// I need to fix the broken Python course first.
// It seems I cut it after "Low Level"
// I will try to restore it if possible by looking for the markers.
// Actually, it's easier to just fix the whole logic.

function processFile(data) {
    // Find all occurrences of test: [
    let result = '';
    let lastIndex = 0;
    const testStartMarker = 'test: [';
    
    let index = data.indexOf(testStartMarker);
    while (index !== -1) {
        result += data.substring(lastIndex, index + testStartMarker.length);
        
        // Now we are inside the array. We need to find the closing ]
        // But we want to limit to 10 questions.
        // Questions are objects { ... }
        
        let braceCount = 0;
        let bracketCount = 1; // We started at [
        let questions = [];
        let currentQuestionStart = -1;
        let i = index + testStartMarker.length;
        
        while (i < data.length && (bracketCount > 0)) {
            const char = data[i];
            if (char === '{' && braceCount === 0) {
                if (currentQuestionStart === -1) currentQuestionStart = i;
                braceCount++;
            } else if (char === '{') {
                braceCount++;
            } else if (char === '}') {
                braceCount--;
                if (braceCount === 0 && currentQuestionStart !== -1) {
                    questions.push(data.substring(currentQuestionStart, i + 1));
                    currentQuestionStart = -1;
                }
            } else if (char === '[' && braceCount === 0) {
                bracketCount++;
            } else if (char === ']' && braceCount === 0) {
                bracketCount--;
                if (bracketCount === 0) {
                    // We found the end of the test array
                    // Keep only 10 questions
                    const limited = questions.slice(0, 10);
                    result += '\n      ' + limited.join(',\n      ') + '\n    ';
                    lastIndex = i;
                    break;
                }
            }
            i++;
        }
        
        index = data.indexOf(testStartMarker, i);
    }
    result += data.substring(lastIndex);
    return result;
}

// Note: If the file is broken (missing brackets), the above might fail or skip.
// Let's try to fix known double comma issues AND limit.
const fixedContent = processFile(content);

// Final cleanup: fix any double commas or empty items that might have been introduced or existed
const finalContent = fixedContent.replace(/\},,\s*/g, '},\n      ');

fs.writeFileSync(filePath, finalContent, 'utf8');
console.log('Successfully limited questions to 10 and fixed structure.');
